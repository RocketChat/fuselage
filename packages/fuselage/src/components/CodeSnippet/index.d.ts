import { ComponentProps, ReactElement } from 'react';

import { Box } from '../Box';

type CodeSnippetProps = ComponentProps<typeof Box> & {
  code?: string;
  buttonText?: string;
  handleClick?: () => void;
};
export const CodeSnippet: (props: CodeSnippetProps) => ReactElement;
