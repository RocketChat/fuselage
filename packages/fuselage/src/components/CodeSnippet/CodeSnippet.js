import React from 'react';

import { Box } from '../Box';
import { Button } from '../Button';

const CodeSnippet = ({ code, handleClick, buttonText = 'Copy', ...props }) => (
  <Box is='pre' rcx-code-snippet {...props}>
    <Box rcx-code-snippet__codebox>
      <code>{code}</code>
    </Box>
    <Box>
      <Button small primary onClick={handleClick}>
        {buttonText}
      </Button>
    </Box>
  </Box>
);

export default CodeSnippet;
