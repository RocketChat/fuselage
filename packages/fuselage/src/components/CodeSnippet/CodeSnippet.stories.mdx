import { action } from '@storybook/addon-actions';
import { Meta, Canvas, ArgsTable, Story } from '@storybook/addon-docs/blocks';

import { CodeSnippet } from '../..';

<Meta
  title='Data Display/CodeSnippet'
  parameters={{ jest: ['CodeSnippet/spec'] }}
/>

# CodeSnippet

The `CodeSnippet` is used to show code or commands and make easier to copy them.

The default button text is `Copy` but you can use the `buttonText` prop to handle translations in your project.

Please check the `useClipBoard` hook in `fuselage-hooks` package, to handle the copy behaviour.

<Canvas>
  <Story name='Default'>
    <CodeSnippet
      code='curl -L https://go.rocket.chat/i/docker-compose.yml -O'
      handleClick={() => {
        action('click');
      }}
    />
  </Story>
</Canvas>

<ArgsTable of={CodeSnippet} />
