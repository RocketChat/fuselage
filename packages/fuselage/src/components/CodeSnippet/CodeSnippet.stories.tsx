import { action } from '@storybook/addon-actions';
import type { ComponentStory, ComponentMeta } from '@storybook/react';
import React from 'react';

import { CodeSnippet } from '../..';

export default {
  title: 'Data Display/CodeSnippet',
  component: CodeSnippet,
  parameters: {
    docs: {
      description: {
        component: `The \`CodeSnippet\` is used to show code or commands and make easier to copy them. \n\n The default button text is \`Copy\` but you can use the \`buttonText\` prop to handle translations in your project. \n\n Please check the \`useClipBoard\` hook in \`fuselage-hooks\` package, to handle the copy behaviour.
				`,
      },
    },
  },
} as ComponentMeta<typeof CodeSnippet>;

const Template: ComponentStory<typeof CodeSnippet> = (args) => (
  <CodeSnippet {...args} />
);

export const Default = Template.bind({});
Default.args = {
  children: 'curl -L https://go.rocket.chat/i/docker-compose.yml -O',
  onClick: action('click'),
};

export const NoCopyButton = Template.bind({});
NoCopyButton.args = {
  children: 'curl -L https://go.rocket.chat/i/docker-compose.yml -O',
};

export const LoadingCode = Template.bind({});
LoadingCode.args = {
  children: '',
  onClick: action('click'),
};
