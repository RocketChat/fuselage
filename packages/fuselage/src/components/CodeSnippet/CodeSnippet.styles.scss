@use '../../styles/colors.scss';
@use '../../styles/lengths.scss';
@use '../../styles/typography.scss';

.rcx-code-snippet {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;

  width: 100%;

  min-height: lengths.size(60);
  padding: lengths.padding(16);

  color: colors.font(default);

  border-radius: theme(
    'code-snippet-border-radius',
    lengths.border-radius(medium)
  );
  background-color: colors.surface(neutral);

  &__codebox {
    margin-right: lengths.margin(8);

    white-space: pre-line;
    word-break: break-all;

    font-family: monospace;
  }
}
