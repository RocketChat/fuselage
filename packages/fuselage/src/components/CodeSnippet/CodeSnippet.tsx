import type { ComponentProps, ReactElement } from 'react';
import React from 'react';

import Box from '../Box';
import { Button } from '../Button';
import { Skeleton } from '../Skeleton';

type CodeSnippetProps = ComponentProps<typeof Box> & {
  children: string;
  buttonText?: string;
  onClick?: () => void;
};

const CodeSnippet = ({
  children,
  onClick,
  buttonText = 'Copy',
  ...props
}: CodeSnippetProps): ReactElement<CodeSnippetProps> => {
  if (!children) {
    return (
      <Box is='pre' rcx-code-snippet {...props}>
        <Skeleton w='100%' />
      </Box>
    );
  }

  return (
    <Box is='pre' rcx-code-snippet {...props}>
      <Box rcx-code-snippet__codebox>
        <code>{children}</code>
      </Box>
      {onClick && children && (
        <Box>
          <Button small primary onClick={onClick}>
            {buttonText}
          </Button>
        </Box>
      )}
    </Box>
  );
};

export default CodeSnippet;
