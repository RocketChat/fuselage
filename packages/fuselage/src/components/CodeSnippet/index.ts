export {
  default,
  default as CodeSnippet,
  CodeSnippetProps,
} from './CodeSnippet';
