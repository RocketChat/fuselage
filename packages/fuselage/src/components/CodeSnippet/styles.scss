@use '../../styles/colors.scss';
@use '../../styles/lengths.scss';
@use '../../styles/typography.scss';

.rcx-code-snippet {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;

  width: 100%;

  min-height: lengths.size(60);
  padding: lengths.padding(16);

  border-radius: lengths.border-width(2);
  background-color: colors.neutral(300);

  &__codebox {
    margin-right: lengths.margin(8);

    white-space: pre-line;
    word-break: break-all;
  }
}
