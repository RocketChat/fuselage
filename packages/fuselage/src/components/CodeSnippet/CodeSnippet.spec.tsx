import { render } from '@testing-library/react';
import React from 'react';

import CodeSnippet from './CodeSnippet';

it('renders without crashing', () => {
  render(<CodeSnippet onClick={() => undefined}>code</CodeSnippet>);
});
