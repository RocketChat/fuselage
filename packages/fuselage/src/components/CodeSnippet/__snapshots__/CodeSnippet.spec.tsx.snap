// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[CodeSnippet Rendering] renders CopyButton without crashing 1`] = `
<body>
  <div>
    <pre
      class="rcx-box rcx-box--full rcx-code-snippet"
    >
      <div
        class="rcx-box rcx-box--full rcx-code-snippet__codebox"
        role="code"
      >
        curl -L https://go.rocket.chat/i/docker-compose.yml -O
      </div>
      <div
        class="rcx-box rcx-box--full"
      >
        <button
          class="rcx-box rcx-box--full rcx-button--small rcx-button--primary rcx-button"
          type="button"
        >
          <span
            class="rcx-button--content"
          >
            Copy
          </span>
        </button>
      </div>
    </pre>
  </div>
</body>
`;

exports[`[CodeSnippet Rendering] renders CustomButtonName without crashing 1`] = `
<body>
  <div>
    <pre
      class="rcx-box rcx-box--full rcx-code-snippet"
    >
      <div
        class="rcx-box rcx-box--full rcx-code-snippet__codebox"
        role="code"
      >
        curl -L https://go.rocket.chat/i/docker-compose.yml -O
      </div>
      <div
        class="rcx-box rcx-box--full"
      >
        <button
          class="rcx-box rcx-box--full rcx-button--small rcx-button--primary rcx-button"
          type="button"
        >
          <span
            class="rcx-button--content"
          >
            Custom name
          </span>
        </button>
      </div>
    </pre>
  </div>
</body>
`;

exports[`[CodeSnippet Rendering] renders Default without crashing 1`] = `
<body>
  <div>
    <pre
      class="rcx-box rcx-box--full rcx-code-snippet"
    >
      <div
        class="rcx-box rcx-box--full rcx-code-snippet__codebox"
        role="code"
      >
        curl -L https://go.rocket.chat/i/docker-compose.yml -O
      </div>
    </pre>
  </div>
</body>
`;

exports[`[CodeSnippet Rendering] renders DisabledButton without crashing 1`] = `
<body>
  <div>
    <pre
      class="rcx-box rcx-box--full rcx-code-snippet"
    >
      <div
        class="rcx-box rcx-box--full rcx-code-snippet__codebox"
        role="code"
      >
        curl -L https://go.rocket.chat/i/docker-compose.yml -O
      </div>
      <div
        class="rcx-box rcx-box--full"
      >
        <button
          class="rcx-box rcx-box--full rcx-button--small rcx-button--primary rcx-button"
          disabled=""
          type="button"
        >
          <span
            class="rcx-button--content"
          >
            Disabled Button
          </span>
        </button>
      </div>
    </pre>
  </div>
</body>
`;

exports[`[CodeSnippet Rendering] renders LoadingCode without crashing 1`] = `
<body>
  <div>
    <pre
      class="rcx-box rcx-box--full rcx-code-snippet"
    >
      <span
        aria-busy="true"
        aria-hidden="true"
        class="rcx-skeleton rcx-skeleton--text rcx-css-1qcz93u"
      />
    </pre>
  </div>
</body>
`;
